### PR TITLE
AbstractMatrix{T}(::SpecialMat{T}) should make a copy

### PR DESCRIPTION
**Pull request by dlfivefifty**
This behaviour seems wrong:
```julia
julia> A = randn(3,3);

julia> AbstractMatrix{Float64}(Symmetric(A))[1,1] = 2;

julia> A
3×3 Matrix{Float64}:
 2.0       0.459786  -0.244634
 0.1644    1.6198    -0.274723
 1.15957  -0.660809  -1.85305
```
If agreed I can add some tests for the fixed behaviour.

### CODE REVIEW

**Comment by dkarrasch** [comment]
I think this is consistent with the standard case

```julia
julia> A = randn(3,3);

julia> AbstractMatrix{Float64}(A)[1,1] = 2;

julia> A
3×3 Matrix{Float64}:
 -0.436256  -0.918073  -0.0848438
 -1.00618   -1.29911   -0.428438
  1.05075    0.370571   1.58952
```

so I agree.

**Comment by dkarrasch** [comment]
Should we make corresponding changes to other AbstractMatrix subtypes? I'll perhaps do that and then we can still revert or close the PR if people disagree.

**Comment by dlfivefifty** [comment]
Absolutely! I'm unfortunately a bit busy right now so any help would be greatly appreciated.

**Review by dkarrasch — COMMENTED** [review]
The rationale behind this PR is the following IIUC. Those who only care about the result type should use `convert`. Those who care about aliasing should be able to rely on constructor-like calls like `AbstractMatrix{T}(A)` or `Bidiagonal{T)(::Bidiagonal)` to produce a copy of the argument. The only exception from the latter are view types (like `*triangular` and `HermOrSym`), which should not copy if the view types match (EDIT: _if called without a target eltype!_).

Does that sound like a reasonable summary of what we're trying to achieve here, @dlfivefifty ? Just so that reviewers can form an opinion more easily.

**Comment by dlfivefifty** [comment]
> The only exception from the latter are view types (like *triangular and HermOrSym), which should not copy if the view types match

One option (which unfortunately may count as breaking) is for eg `Symmetric{T}(A)` to always make a copy whilst `Symmetric(A)` is a lazy view. This would then be consistent...

**Comment by dkarrasch** [comment]
You mean `Symmetric{T}(A::AbstractMatrix)` to always make a copy? Even for `A::Matrix`? In that special case, that is actually doable and non-breaking, because currently there is no such constructor:

```julia
julia> using LinearAlgebra

julia> Symmetric{Float64}(zeros(3,3))
ERROR: MethodError: no method matching (Symmetric{Float64, S} where S<:(AbstractMatrix{<:Float64}))(::Matrix{Float64})
Stacktrace:
 [1] top-level scope
   @ REPL[2]:1

julia> Symmetric(zeros(3,3))
3×3 Symmetric{Float64, Matrix{Float64}}:
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0
```

For the triangular family, such a constructor exists, and in this PR I made it always create a copy, even in the matching eltype case. One has to keep in mind that the question about whether or not to make a copy only affects the matching eltype case. It sounds strange to me to rely on non-copying behavior in that single special case. And again, if one wants that optimization, there is always the `convert` option.

**Comment by dkarrasch** [comment]
> So it sounds like we are in agreement: if you specify the eltype than it should always make a copy, regardless of whether eltype matches. And if you don't specify the eltype then it doesn't make the copy.

It depends on what you think about the `*diagonal` types, which are not views. You proposed to make them copy even without a specification of the target eltype in the initial commit, like `Bidiagonal(B::Bidiagonal) = copy(B)`. I'd argue that makes a lot of sense because `Bidiagonal(A)` is not a bidiagonal _view_ of `A`, but a bidiagonal _copy_ of `A`. One could extend the argument and say that `Bidiagonal(B::Bidiagonal)` should then also be a bidiagonal copy of `B`.

**Comment by dkarrasch** [comment]
> If `Bidiagonal(::Bidiagonal)` makes a copy then the behaviour differs depending on the input.

That's the current behaviour.

```julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.0-beta2 (2023-08-17)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using LinearAlgebra

julia> function foo!(A::AbstractMatrix)
           B = Tridiagonal(A)
           B[2,1] = 3
           B
       end
foo! (generic function with 1 method)

julia> A = zeros(3,3);

julia> foo!(A)
3×3 Tridiagonal{Float64, Vector{Float64}}:
 0.0  0.0   ⋅ 
 3.0  0.0  0.0
  ⋅   0.0  0.0

julia> A
3×3 Matrix{Float64}:
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

julia> B = Tridiagonal(zeros(2), zeros(3), zeros(2));

julia> foo!(B)
3×3 Tridiagonal{Float64, Vector{Float64}}:
 0.0  0.0   ⋅ 
 3.0  0.0  0.0
  ⋅   0.0  0.0

julia> B
3×3 Tridiagonal{Float64, Vector{Float64}}:
 0.0  0.0   ⋅ 
 3.0  0.0  0.0
  ⋅   0.0  0.0
```

I picked `Tridiagonal` because `Bidiagonal(A, ::Symbol)` _always_ copies the data, independent from whether `A` is bidiagonal or not. That's exactly my point: the `*diagonal` family are not views, they genericly read out the diagonals, freshly allocated, and construct the `*diagonal` object based on that... well, except when we happen to have the same matrix type. Perhaps, we should actually simply remove the specialized methods and leave it to `diag(A, k)`, which do create copies of `T.dv` etc.

File: stdlib/LinearAlgebra/src/bidiag.jl [hunk 1]
```diff
@@ -22,9 +22,9 @@ function Bidiagonal{T}(dv::AbstractVector, ev::AbstractVector, uplo::Union{Symbo
                convert(AbstractVector{T}, ev)::AbstractVector{T},
                uplo)
 end
-function Bidiagonal{T,V}(A::Bidiagonal) where {T,V<:AbstractVector{T}}
-    Bidiagonal{T,V}(A.dv, A.ev, A.uplo)
-end
+# function Bidiagonal{T,V}(A::Bidiagonal) where {T,V<:AbstractVector{T}}
+#     Bidiagonal{T,V}(A.dv, A.ev, A.uplo)
+# end
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/bidiag.jl:27]: The corresponding method for `A::AbstractMatrix` doesn't even exist! The same applies to `[Sym]Tridiagonal`. That's not generic programming, right? If you specify `T` and `V`, you already know that you're working with a, for instance, `Bidiagonal` here.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/bidiag.jl:27]: Ok, I realized these are actually needed for the `convert` methods, so perhaps rather internal.

**Comment by dkarrasch** [comment]
After a long while I realized the issue was that `AbstractMatrix{T}(::LinAlgMatrix{T})` did not copy, because it was falling back on, say, `SymTridiagonal(SymTridiagonal)`. I think a common reason to end up calling `AbstractMatrix{T}(A)` is due to calling `convert(AbstractMatrix{T}, A)`. The latter checks if `A isa AbstractMatrix{T}`, which fails only when the eltype of `A` doesn't match `T`. [If they match, `A` is returned directly.] In that case, we call `AbstractMatrix{T}(A)`, which needs to create a copy anyway. That means, we only need to fix the case when somebody calls `AbstractMatrix{T}(A)` explicitly and assumes that the result does not alias `A`. So, we can keep the non-copying behaviour of `Diagonal(::Diagonal)` and friends and keep the changes non-breaking.

**Pull request merged**